api/src/wfl/module/covid.clj

```diff
@@ -183,17 +183,120 @@
+      (if (= job-status "succeeded")
+        (assoc result :snapshot_id (:id (datarepo/get-job-result id)))
+        (assoc result :snapshot_id nil)))))
+
+(defn ^:private write-snapshot-id
+  "Write `snapshot_id` and `job_status` into source `details` table
+   from the `_tdr-job-metadata` map, update timestamp with real now."
+  [{:keys [details] :as _source}
+   {:keys [id job_status snapshot_id] :as _tdr-job-metadata}]
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/update! tx details {:snapshot_creation_job_status job_status
+                              :snapshot_id snapshot_id
+                              :updated  (OffsetDateTime/now)}
+      ;; we don't have the primary key reference of details table here,
+      ;; so use job-id for where clause, since job-id is unique in TDR,
+      ;; this shall be fine
+                  ["snapshot_creation_job_id = ?" id])))
+
+(defn ^:private write-snapshots-creation-jobs
+  "Write all `snapshots-creation-jobs` along with the `row-ids` they
+   try to snapshot into source `details` table, with the frozen `now`."
+  [{:keys [last_checked details] :as _source} now row-ids snapshots-creation-jobs]
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/insert-multi! tx
+                        details
+                        (mapv (fn [id] {:snapshot_creation_job_id id
+                                        :datarepo_row_ids         row-ids
+                                        :start_time               last_checked
+                                        :end_time                 now})
+                              snapshots-creation-jobs))))
+
+(defn ^:private update-last-checked
+  "Update the `last_checked` field in source table with
+   the frozen `now`."
+  [{:keys [id] :as _source} now]
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/update! tx tdr-source-table {:last_checked now} ["id = ?" id])))
+
 ;; Create and add new snapshots to the snapshot queue
-(defn ^:private update-tdr-source [source]
-  (letfn [(find-new-rows       [source now] [])
-          (make-snapshots      [source row-ids])
-          (write-snapshots     [source snapshots])
-          (update-last-checked [source now])]
-    (let [now     (OffsetDateTime/now)
-          row-ids (find-new-rows source now)]
-      (when-not (empty? row-ids)
-        (write-snapshots source (make-snapshots source row-ids)))
-      (update-last-checked source now))))
+(defn ^:private update-tdr-source
+  "Check for new data in TDR, create new snapshots, insert
+   resulting job creation ids into database and update the
+   timestamp for next time using transaction TX."
+  [source]
+  ;; attempt to snapshot new rows in TDR
+  (let [utc-now-obj     (OffsetDateTime/now (ZoneId/of "UTC"))
+        utc-now         (.format utc-now-obj (DateTimeFormatter/ofPattern "yyyy-MM-dd HH:mm:ss"))
+        row-ids         (vec (find-new-rows source utc-now))
```

Comment:
As `: datarepo_row_ids` is a `TEXT[]` in DB, the postgres (or JDBC) wants something other than a lazy seq, otherwise it throws. Same reason for why I had to use `doall` in another place. Given what we discussed in https://github.com/broadinstitute/wfl/pull/371#discussion_r633655546, I have moved the `vec` here to the actual place that requires it.

Reply:
updated